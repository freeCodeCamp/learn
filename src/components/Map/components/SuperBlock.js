import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import uniq from 'lodash/uniq';
import find from 'lodash/find';

import Block from './Block';

import { makeExpandedSuperBlockSelector, toggleSuperBlock } from '../redux';
import Caret from '../../icons/Caret';
import { ChallengeNode } from '../../../redux/propTypes';

const mapStateToProps = (state, ownProps) => {
  const expandedSelector = makeExpandedSuperBlockSelector(ownProps.superBlock);

  return createSelector(expandedSelector, isExpanded => ({ isExpanded }))(
    state
  );
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      toggleSuperBlock
    },
    dispatch
  );
}

const propTypes = {
  introNodes: PropTypes.arrayOf(
    PropTypes.shape({
      fields: PropTypes.shape({ slug: PropTypes.string.isRequired }),
      frontmatter: PropTypes.shape({
        title: PropTypes.string.isRequired,
        block: PropTypes.string.isRequired
      })
    })
  ),
  isExpanded: PropTypes.bool,
  nodes: PropTypes.arrayOf(ChallengeNode),
  superBlock: PropTypes.string,
  toggleSuperBlock: PropTypes.func.isRequired
};

const codingPrepRE = new RegExp('Interview Prep');

function createSuperBlockTitle(str) {
  return codingPrepRE.test(str)
    ? `${str} (Thousands of hours of challenges)`
    : `${str} Certification (300 hours)`;
}

export class SuperBlock extends PureComponent {
  renderBlock(superBlock) {
    const { nodes, introNodes } = this.props;
    const blocksForSuperBlock = nodes.filter(
      node => node.superBlock === superBlock
    );
    const blockDashedNames = uniq(
      blocksForSuperBlock.map(({ block }) => block)
    );
    return (
      <ul>
        {blockDashedNames.map(blockDashedName => (
          <Block
            blockDashedName={blockDashedName}
            challenges={blocksForSuperBlock.filter(
              node => node.block === blockDashedName
            )}
            intro={find(
              introNodes,
              ({ frontmatter: { block } }) =>
                block
                  .toLowerCase()
                  .split(' ')
                  .join('-') === blockDashedName
            )}
            key={blockDashedName}
          />
        ))}
      </ul>
    );
  }

  render() {
    const { superBlock, isExpanded, toggleSuperBlock } = this.props;
    return (
      <li className={`superblock ${isExpanded ? 'open' : ''}`}>
        <div className='map-title' onClick={() => toggleSuperBlock(superBlock)}>
          <Caret />
          <h4>{createSuperBlockTitle(superBlock)}</h4>
        </div>
        {isExpanded ? this.renderBlock(superBlock) : null}
      </li>
    );
  }
}

SuperBlock.displayName = 'SuperBlock';
SuperBlock.propTypes = propTypes;

export default connect(mapStateToProps, mapDispatchToProps)(SuperBlock);
