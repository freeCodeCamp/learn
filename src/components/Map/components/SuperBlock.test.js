/* global expect */

import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';

import { SuperBlock } from './SuperBlock';
import mockNodes from '../../../__mocks__/map-nodes';
import mockIntroNodes from '../../../__mocks__/intro-nodes';

Enzyme.configure({ adapter: new Adapter() });
const renderer = new ShallowRenderer();

test('<SuperBlock /> not expanded snapshot', () => {
  const toggleSpy = sinon.spy();
  const props = {
    introNodes: mockIntroNodes,
    isExpanded: false,
    nodes: mockNodes,
    superBlock: 'Super Block One',
    toggleSuperBlock: toggleSpy
  };
  const componentToRender = <SuperBlock {...props} />;
  const component = renderer.render(componentToRender);

  expect(component).toMatchSnapshot('superBlock-not-expanded');
});

test('<SuperBlock /> expanded snapshot', () => {
  const toggleSpy = sinon.spy();
  const props = {
    introNodes: mockIntroNodes,
    isExpanded: true,
    nodes: mockNodes,
    superBlock: 'Super Block One',
    toggleSuperBlock: toggleSpy
  };
  const componentToRender = <SuperBlock {...props} />;
  const component = renderer.render(componentToRender);

  expect(component).toMatchSnapshot('superBlock-expanded');
});

test('<SuperBlock should handle toggle clicks correctly', () => {
  const toggleSpy = sinon.spy();
  const props = {
    introNodes: mockIntroNodes,
    isExpanded: false,
    nodes: mockNodes,
    superBlock: 'Super Block One',
    toggleSuperBlock: toggleSpy
  };
  const componentToRender = <SuperBlock {...props} />;
  const enzymeWrapper = Enzyme.shallow(componentToRender);

  expect(toggleSpy.called).toBe(false);
  expect(
    enzymeWrapper
      .find('.map-title')
      .find('h4')
      .text()
  ).toBe('Super Block One Certification (300 hours)');
  expect(enzymeWrapper.find('ul').length).toBe(0);

  enzymeWrapper.find('.map-title').simulate('click');

  expect(toggleSpy.called).toBe(true);
  expect(toggleSpy.calledWithExactly('Super Block One')).toBe(true);

  enzymeWrapper.setProps({ ...props, isExpanded: true });

  expect(
    enzymeWrapper
      .find('.map-title')
      .find('h4')
      .text()
  ).toBe('Super Block One Certification (300 hours)');
  expect(enzymeWrapper.find('ul').length).toBe(1);
});
