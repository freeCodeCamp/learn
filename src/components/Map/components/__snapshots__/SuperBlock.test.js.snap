// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SuperBlock /> expanded snapshot: superBlock-expanded 1`] = `
<li
  className="superblock open"
>
  <div
    className="map-title"
    onClick={[Function]}
  >
    <Caret />
    <h4>
      Super Block One Certification (300 hours)
    </h4>
  </div>
  <ul>
    <Connect(Block)
      blockDashedName="block-a"
      challenges={
        Array [
          Object {
            "block": "block-a",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-one",
            },
            "id": "a",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge One",
          },
          Object {
            "block": "block-a",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-two",
            },
            "id": "b",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge Two",
          },
        ]
      }
      intro={
        Object {
          "fields": Object {
            "slug": "/super-block-one/block-a",
          },
          "frontmatter": Object {
            "block": "Block A",
            "title": "Introduction to Block A",
          },
        }
      }
    />
    <Connect(Block)
      blockDashedName="block-b"
      challenges={
        Array [
          Object {
            "block": "block-b",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-one",
            },
            "id": "c",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge One",
          },
          Object {
            "block": "block-b",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-two",
            },
            "id": "d",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge Two",
          },
        ]
      }
      intro={
        Object {
          "fields": Object {
            "slug": "/super-block-one/block-b",
          },
          "frontmatter": Object {
            "block": "Block B",
            "title": "Introduction to Block B",
          },
        }
      }
    />
    <Connect(Block)
      blockDashedName="block-c"
      challenges={
        Array [
          Object {
            "block": "block-c",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block C",
              "slug": "/super-block-one/block-c/challenge-one",
            },
            "id": "e",
            "isPrivate": true,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge One",
          },
        ]
      }
      intro={
        Object {
          "fields": Object {
            "slug": "/super-block-one/block-c",
          },
          "frontmatter": Object {
            "block": "Block C",
            "title": "Introduction to Block C",
          },
        }
      }
    />
  </ul>
</li>
`;

exports[`<SuperBlock /> not expanded snapshot: superBlock-not-expanded 1`] = `
<li
  className="superblock "
>
  <div
    className="map-title"
    onClick={[Function]}
  >
    <Caret />
    <h4>
      Super Block One Certification (300 hours)
    </h4>
  </div>
</li>
`;
